tx_check.go

```diff
@@ -0,0 +1,202 @@
+		for i := pgid(0); i <= pgid(p.overflow); i++ {
+			var id = p.id + i
+			if _, ok := reachable[id]; ok {
+				ch <- fmt.Errorf("page %d: multiple references (stack: %v)", int(id), stack)
+			}
+			reachable[id] = p
+		}
+
+		// We should only encounter un-freed leaf and branch pages.
+		if freed[p.id] {
+			ch <- fmt.Errorf("page %d: reachable freed", int(p.id))
+		} else if (p.flags&branchPageFlag) == 0 && (p.flags&leafPageFlag) == 0 {
+			ch <- fmt.Errorf("page %d: invalid type: %s (stack: %v)", int(p.id), p.typ(), stack)
+		}
+	})
+
+	tx.recursivelyCheckPages(b.root, keyValueStringer.KeyToString, ch)
+
+	// Check each bucket within this bucket.
+	_ = b.ForEachBucket(func(k []byte) error {
+		if child := b.Bucket(k); child != nil {
+			tx.checkBucket(child, reachable, freed, keyValueStringer, ch)
+		}
+		return nil
+	})
+}
+
+// recursivelyCheckPages confirms database consistency with respect to b-tree
+// key order constraints:
+//   - keys on pages must be sorted
+//   - keys on children pages are between 2 consecutive keys on the parent's branch page).
+func (tx *Tx) recursivelyCheckPages(pgid pgid, keyToString func([]byte) string, ch chan error) {
+	tx.recursivelyCheckPagesInternal(pgid, nil, nil, nil, keyToString, ch)
+}
+
+// recursivelyCheckPagesInternal verifies that all keys in the subtree rooted at `pgid` are:
+//   - >=`minKeyClosed` (can be nil)
+//   - <`maxKeyOpen` (can be nil)
+//   - Are in right ordering relationship to their parents.
+//     `pagesStack` is expected to contain IDs of pages from the tree root to `pgid` for the clean debugging message.
+func (tx *Tx) recursivelyCheckPagesInternal(
+	pgid pgid, minKeyClosed, maxKeyOpen []byte, pagesStack []pgid,
+	keyToString func([]byte) string, ch chan error) (maxKeyInSubtree []byte) {
+
+	p := tx.page(pgid)
+	pagesStack = append(pagesStack, pgid)
+	switch {
+	case p.flags&branchPageFlag != 0:
+		// For branch page we navigate ranges of all subpages.
+		runningMin := minKeyClosed
+		for i := range p.branchPageElements() {
+			elem := p.branchPageElement(uint16(i))
+			if i == 0 && runningMin != nil && compareKeys(runningMin, elem.key()) > 0 {
+				ch <- fmt.Errorf("key (%d, %s) on the branch page(%d) needs to be >="+
+					" to the index in the ancestor. Pages stack: %v",
+					i, keyToString(elem.key()), pgid, pagesStack)
+			}
+
+			if maxKeyOpen != nil && compareKeys(elem.key(), maxKeyOpen) >= 0 {
+				ch <- fmt.Errorf("key (%d: %s) on the branch page(%d) needs to be <"+
+					" than key of the next element reachable from the ancestor (%v). Pages stack: %v",
+					i, keyToString(elem.key()), pgid, keyToString(maxKeyOpen), pagesStack)
+			}
+
+			var maxKey []byte
+			if i < len(p.branchPageElements())-1 {
+				maxKey = p.branchPageElement(uint16(i + 1)).key()
+			} else {
+				maxKey = maxKeyOpen
+			}
+			maxKeyInSubtree = tx.recursivelyCheckPagesInternal(elem.pgid, elem.key(), maxKey, pagesStack, keyToString, ch)
+			runningMin = maxKeyInSubtree
+		}
+		return
```

Comment:
```suggestion
		return maxKeyInSubtree
```

Reply:
Why not `return maxKeyInSubtree` that is implicit ?

Reply:
Right, we should return `maxKeyInSubtree`. For the rightest element in a branch page, the maxKey in the page it links to should be the largest key in the tree rooted from current branch page.